internal/worker/concurrency_auto_scaler_test.go

```diff
@@ -0,0 +1,390 @@
+				{common.PtrOf(false), common.PtrOf(int64(1000))}, // <- disable
+			},
+			[]eventLog{
+				{autoScalerEventStart, false, 100, "00:00:00"},
+				{autoScalerEventEnable, true, 100, "00:00:00"},
+				{autoScalerEventPollerSkipUpdateCooldown, true, 100, "00:00:01"},
+				{autoScalerEventPollerScaleDown, true, 80, "00:00:02"},
+				{autoScalerEventDisable, false, 100, "00:00:02"},
+				{autoScalerEventPollerSkipUpdateNotEnabled, false, 100, "00:00:03"},
+				{autoScalerEventStop, false, 100, "00:00:03"},
+			},
+		},
+		{
+			"idle pollers and enabled at a later time",
+			[]*shared.AutoConfigHint{
+				{common.PtrOf(false), common.PtrOf(int64(1000))}, // <- tick, in cool down
+				{common.PtrOf(false), common.PtrOf(int64(1000))}, // <- tick, not enabled
+				{common.PtrOf(true), common.PtrOf(int64(1000))},  // <- tick, enable scale up
+			},
+			[]eventLog{
+				{autoScalerEventStart, false, 100, "00:00:00"},
+				{autoScalerEventPollerSkipUpdateNotEnabled, false, 100, "00:00:01"},
+				{autoScalerEventPollerSkipUpdateNotEnabled, false, 100, "00:00:02"},
+				{autoScalerEventEnable, true, 100, "00:00:02"},
+				{autoScalerEventPollerScaleDown, true, 80, "00:00:03"},
+				{autoScalerEventStop, true, 80, "00:00:03"},
+			},
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			defer goleak.VerifyNone(t)
+			core, obs := observer.New(zap.DebugLevel)
+			logger := zap.New(core)
+			clock := clockwork.NewFakeClockAt(time.Date(2024, 1, 1, 0, 0, 0, 0, time.UTC))
+			scaler := createTestConcurrencyAutoScaler(t, logger, clock)
+
+			// mock poller every 1 tick time
+			var wg sync.WaitGroup
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				clock.Sleep(testTickTime / 2) // poll delay by 0.5 unit of time to avoid test flakiness
+				for _, hint := range tt.pollAutoConfigHint {
+					t.Log("hint process time: ", clock.Now().Format(testTimeFormat))
+					for i := 0; i < numberOfPollsInRollingAverage; i++ { // simulate polling multiple times to avoid flakiness
+						scaler.ProcessPollerHint(hint)
+					}
+					clock.Sleep(testTickTime)
+				}
+			}()
+
+			scaler.Start()
+			clock.BlockUntil(2)
+
+			// advance clock
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				for i := 0; i < len(tt.pollAutoConfigHint)*2+1; i++ {
+					clock.Advance(testTickTime / 2)
+					time.Sleep(10 * time.Millisecond) // process non-time logic
+				}
+			}()
+
+			wg.Wait()
+			scaler.Stop()
+
+			var actualEvents []eventLog
+			for _, event := range obs.FilterMessage(autoScalerEventLogMsg).All() {
+				if event.ContextMap()["event"] != string(autoScalerEventEmitMetrics) {
+					t.Log("event: ", event.ContextMap())
+					actualEvents = append(actualEvents, eventLog{
+						eventType:   autoScalerEvent(event.ContextMap()["event"].(string)),
+						enabled:     event.ContextMap()["enabled"].(bool),
+						pollerQuota: event.ContextMap()["poller_quota"].(int64),
+						time:        event.ContextMap()["time"].(time.Time).Format(testTimeFormat),
```

Comment:
```suggestion
						time:        event.Time.Format(testTimeFormat),
```
with a `logger.WithClock`, I think this handles the "logs are hard to identify uniquely" thing that seems to be the intent here?

Reply:
Yes, I just didn't find a good way to assert event logs and find this is easier (or quicker) with a test entity.

Reply:
yea, the "make a simplified struct for comparison" is a good choice I think, this was just for the time-context-map-field.

Reply:
I see. I'll remove the time field then.